pkg/util/docker/event_stream.go

```diff
@@ -19,63 +18,89 @@ import (
 	"github.com/docker/docker/api/types/filters"
 )
 
-//// Can't be unit tested, covered by the listeners/docker
-//// and dogstatsd/origin_detection integration tests.
+//// eventStreamState logic unit tested in event_stream_test.go
+//// DockerUtil logic covered by the listeners/docker and dogstatsd/origin_detection integration tests.
 
-// eventSubscriber holds the state for a subscriber
-type eventSubscriber struct {
-	name       string
-	eventChan  chan *ContainerEvent
-	errorChan  chan error
-	cancelChan chan struct{}
-}
+const eventSendTimeout = 100 * time.Millisecond
+const eventSendBuffer = 5
 
 // SubscribeToContainerEvents allows a package to subscribe to events from the event stream.
-// An unique subscriber name should be provided.
+// A unique subscriber name should be provided.
 func (d *DockerUtil) SubscribeToContainerEvents(name string) (<-chan *ContainerEvent, <-chan error, error) {
-	d.RLock()
-	if _, found := d.eventSubscribers[name]; found {
-		d.RUnlock()
-		return nil, nil, errors.New("already subscribed")
+	c1, c2, err, shouldStart := d.eventState.subscribe(name)
+
+	if shouldStart {
+		d.eventState.Lock()
+		d.eventState.running = true
+		go d.dispatchEvents(d.eventState.cancelChan)
+		d.eventState.Unlock()
 	}
-	d.RUnlock()
+
+	return c1, c2, err
+}
+
+func (e *eventStreamState) subscribe(name string) (<-chan *ContainerEvent, <-chan error, error, bool) {
+	var shouldStart bool
+	e.RLock()
+	if _, found := e.subscribers[name]; found {
+		e.RUnlock()
+		return nil, nil, ErrAlreadySubscribed, false
+	}
+	e.RUnlock()
 
 	sub := &eventSubscriber{
-		name:       name,
-		eventChan:  make(chan *ContainerEvent, 5),
-		errorChan:  make(chan error, 5),
-		cancelChan: make(chan struct{}),
+		name:      name,
+		eventChan: make(chan *ContainerEvent, eventSendBuffer),
+		errorChan: make(chan error, 1),
+	}
+	e.Lock()
+	e.subscribers[name] = sub
+	if !e.running {
+		shouldStart = true
 	}
-	d.Lock()
-	d.eventSubscribers[name] = sub
-	d.Unlock()
+	e.Unlock()
 
-	go d.streamEvents(sub.eventChan, sub.cancelChan)
-	return sub.eventChan, sub.errorChan, nil
+	return sub.eventChan, sub.errorChan, nil, shouldStart
 }
 
 // UnsubscribeFromContainerEvents allows a package to unsubscribe.
 // The call is blocking until the request is processed.
 func (d *DockerUtil) UnsubscribeFromContainerEvents(name string) error {
-	d.Lock()
-	sub, found := d.eventSubscribers[name]
+	err, shouldStop := d.eventState.unsubscribe(name)
+
+	if shouldStop {
+		d.eventState.Lock()
+		d.eventState.running = false
+		d.eventState.cancelChan <- struct{}{}
+		d.eventState.Unlock()
+	}
+
+	return err
+}
+
+func (e *eventStreamState) unsubscribe(name string) (error, bool) {
+	var shouldStop bool
+	e.Lock()
+	sub, found := e.subscribers[name]
 	if !found {
-		d.Unlock()
-		return errors.New("not subscribed")
+		e.Unlock()
```

Comment:
if you defer `e.Unlock()` it would avoid writting it twice no?

Reply:
indeed 👍